internal/terraform/rootmodule/root_module.go

```diff
@@ -489,20 +495,42 @@ func (rm *rootModule) parsedFiles() map[string]*hcl.File {
 }
 
 func (rm *rootModule) MergedSchema() (*schema.BodySchema, error) {
+	var err error
+	preloadProviderSchemasOnce.Do(func() {
+		var f http.File
+		f, err = preloadedSchemas.Files.Open("schemas.json")
+		if err != nil {
+			return
+		}
+
+		err = json.NewDecoder(f).Decode(&preloadedProviderSchemas)
+	})
+	if err != nil {
+		rm.logger.Printf("decoding of the preloaded provider schemas has errored and will not be retried: %s", err)
+	}
```

Comment:
How would you feel about extracting this into a method, e.g.

```go
func (rm *rootModule) preloadedProviderSchemas() (*tfjson.ProviderSchemas, error) {
	var err error
	preloadProviderSchemasOnce.Do(func() {
		var f http.File
		f, fErr := preloadedSchemas.Files.Open("schemas.json")
		if err != nil {
			err = fErr
			return
		}

		preloadedProviderSchemas = &tfjson.ProviderSchemas{}
		err = json.NewDecoder(f).Decode(preloadedProviderSchemas)
	})

	return preloadedProviderSchemas, err
}
```
assuming we also turn `preloadedProviderSchemas` variable into a pointer, which probably better reflects that it may be empty - i.e. it can be just `nil` instead of `tfjson.ProviderSchemas{}` with zero values.

Reply:
no problem, I will likely just make it a pure function and not a method of rootModule

Reply:
Yep - that's better in fact as it doesn't need any rootModule data. 👍